### ISSUE TITLE
hover.tooltip convenience function for 'datetime' info

### ISSUE THREAD

**Issue by ijstokes** [bokeh/bokeh#1239]
It would be nice to make it easy to create a "hover" tooltip template that could do some simple conversions/formatting on "standard" data point information: x, y, size, color, etc.  Maybe also some convenience conversions for `source` attributes of data points.

My particular scenario has to do with providing a hover tooltip which displays a human-readable time stamp for the data point.  The x-axis is configured as `datetime`, but when I display `$x` I get back a float (seconds since the epoch) which isn't very readable.  To get around this, I need to add a ColumnDataSource with a single field, like so:

``` python
source = ColumnDataSource(data=dict(
    time=df.map(lambda x: x.strftime('%d-%m-%Y'))
))
```

and then I tell my tooltip to display `("time", "@time")`.  This is workable, but it would be nicer if I could just do:

``` python
("time", "`stftime('%m-%d-%Y', $x)`"
```

or something similar to say "do a client-side conversion of `$x` into something more human-readable"

**Comment by drorata** [comment]
This is a great addition! Thanks! Could someone add some snippets explaining how to use this?

I got so far:

    hover.tooltips = {"x":"$x{int}", "y":"$y{int}"}

will format the values as integers.

Assuming that `x` is a timestamp-integer, it would be really nice to be able to format it into a date-time.

**Comment by Riemann1337** [comment]
I second this feature request.  Including additional formatting options for other datatypes would be nice as well (perhaps use formatting options similar to those that a DataTable has?)

**Comment by Rakeshbart** [comment]
Yes this feature would be really helpful as I have many plots with datetime as the x axis. could you please explain how to use this as well.

**Comment by davideuler** [comment]
I create the DataSource for human readable datetime like this, and then pass the datasource to glyphs.

```
def create_data_source(data_frame):
    return ColumnDataSource(
        data=dict(
            date=[x.strftime("%Y-%m-%d") for x in data_frame['date']],
            open=data_frame['open'],
            close=data_frame['close'],
        )
    )

df_inc = df[df.close > df.open]  # get the subset of data frame whose close is larger than open.
source_inc = create_data_source(df_inc)

p.rect(df.date[inc], mids[inc], w, spans[inc], fill_color="#F2583E", line_color="black",source=source_inc)
```

**Comment by esvhd** [comment]
I also support adding this feature, will be very helpful. My hack right now is to create a column of strings representing the dates in my data frame passed in as source and refer to that column in the tooltip..

**Comment by adriantorrie** [comment]
This feature would be great to have!

Perhaps intially having the [ISO 8601 standard](http://www.iso.org/iso/home/standards/iso8601.htm) covered first with the simple notation of:
* `$x{date-time}` -> output: `'YYYY-MM-DD HH:MM:SS.[fractional seconds]'`
* `$x{date}` -> output: `'YYYY-MM-DD'`
* `$x{time}` -> output: `'HH:MM:SS.[fractional seconds]'`

The actual format of ISO 8601 timestamp has a 'T' or 't' between the date and time parts, e.g ``'YYYY-MM-DDTHH:MM:SS.[fractional seconds]'`, however I've replaced the 'T' with a space character to be human readable. This is acceptable according to [RFC3339 in section 5.6 Internet and Date/Time Format](https://tools.ietf.org/html/rfc3339#section-5.6):

>NOTE: ISO 8601 defines date and time separated by "T".
      Applications using this syntax may choose, for the sake of
      readability, to specify a full-date and full-time separated by
      (say) a space character.

Further examples of international agreed standards are found in [RFC3339](https://tools.ietf.org/html/rfc3339#appendix-A). Note tough that RFC3339 has the following:

>The following is an attempt to create a formal
   grammar from ISO 8601.  This is informational only and may contain
   errors.  ISO 8601 remains the authoritative reference.

The above gives a grammar for format names, some examples for discussion are (their are many more in the RFC document above):
* `date-month` -> usage: `$x{date-month}` -> output: `01-12`
* `date-mday`
* `time-hour`
* `time-offset`

**Comment by AlJohri** [comment]
@mattpap you mentioned the preliminary support for number and time formatting is done - how does one use the time formatting? I only got as far as @drorata with `"$x{int}"`

**Comment by VelizarVESSELINOV** [comment]
👍

**Comment by Rubyj** [comment]
@mattpap Has this functionality been added yet? I have a hover tooltip being outputted as an epoch string when it is a date on the python side.

**Comment by mattpap** [comment]
@Rubyj, no, not yet, though this is something we could add fairly easily as we already ship `timezone` library with bokehjs (used for axis tick label formatting).

**Comment by artur-scholz** [comment]
So this feature (to format unix timestamp as datetime in the hovertool) is requested since 2+ years, and not yet implemented in the latest bokeh release?

The workaround with adding a separate column to the ColumnDataSource where this info is stored as formatted text string is not appealing to me.

Then, why not using the custom tooltip option? For example, this works for my purpose:

```
hover = HoverTool(
            tooltips="""
            <div>
                <p>time_start: <span id="_time_start"></span><p>
                <p>time_end: <span id="_time_end"></span><p>
            </div>
            <script>
            function fun(unix_timestamp) {
                var d1 = new Date(unix_timestamp);
                return d1.toISOString()
            }
            document.getElementById("_time_start").innerHTML = fun(@time_start);
            document.getElementById("_time_end").innerHTML = fun(@time_end);
            </script>
            """
        )
```

Still, I would prefer to have a formatting option in the native tooltip feature, e.g. $x{date-time}

Artur

**Comment by bryevdv** [comment]
> So this feature (to format unix timestamp as datetime in the hovertool) is requested since 2+ years, and not yet implemented in the latest bokeh release?

There is more work to do than is humanly accomplish by the number of people currently available to do it. New contributors are always welcome.

**Comment by Corleo** [comment]
As an improvement suggestion, how about an option to get the own axis formatter to format the hover tooltip so it can be done at the browser side and do the formatting only on part of data that is being displayed in plot?

**Comment by bryevdv** [comment]
I believe this can now fairly easily be accomplished by allowing a `"formatter"` field in data specs, that accept a `Transform`. [*] Then the (optional) transform can be used during any display of the data, e.g. by a hover tool. We can provide some canned format transforms such as for date times, but a `CustomJSTransform` opens up possibilities for users to insert arbitrary transforms. 

[*] Or should it be a tick formatter? the functionality certainly matches the use case but the terminology (naming) is off.

**Comment by bryevdv** [comment]
OK, scaling back to something slightly less ambitious. Will push a PR later today that provide for things like this:
```
p2.add_tools(HoverTool(tooltips=[("date", "@x{%F %T}")], formatters={"x": "datetime"}))
```
Which yields:

<img width="287" alt="screen shot 2017-04-22 at 13 59 50" src="https://cloud.githubusercontent.com/assets/1078448/25307321/a490326a-2764-11e7-910a-1db8627b7746.png">

So `formatters` can optionally specify one of `"datetime"`, `"printf"`, or `"numeral"` for a field name, which will  use `tz`, `SPrintf` or`Numbro` respectively 

Specifying no formatter defaults to `Numbro` which preserves the current behavior

I think this leaves the door open for custom formatter models with something like 
```
HoverTool(tooltips=[("date", "@x{custom}")], formatters={"x": my_custom_formatter})
```
but I don't want to tackle that right now

**Comment by sigurdurb** [comment]
warning and thx to @artur-scholz "to ISO string" method.
It works but it needs to add the {0.0000} formatting for the incoming unix date like so:
```
 document.getElementById("_time_start").innerHTML = fun(@time_start{0.00000000});
 document.getElementById("_time_end").innerHTML = fun(@time_end{0.00000000});
````
I just got the first letters of some unix date rounded to the fourth number like 1493000000000 when I should have got 149291492944633005.431

**Comment by tienlx93** [comment]
Hi @bryevdv ,

I'm currently using HoverTool to format tooltips for x/y axis for datashader trend chart.
However, if I use `@x{%F %T}` format, it only show the daytime of begin of x axis, not current value
![image](https://user-images.githubusercontent.com/9817634/67183083-a6a4d900-f3f1-11e9-9b26-dc9bff69d50d.png)

If I use `$x{%F %T}` format, it only display x in plain format
![image](https://user-images.githubusercontent.com/9817634/67183131-cb00b580-f3f1-11e9-81d8-755cf25c3585.png)
Please help me to explain how to correctly format the `$x` value. Thank you.

Here is the python code:
```python
curve = get_curve(df, 'Legend 1')
curve2 = get_curve(df2, 'Legend 2')
curve_dict = OrderedDict()
curve_dict['Legend 1'] = curve
curve_dict['Legend 2'] = curve2
overlay = hv.NdOverlay(curve_dict, kdims='types']).opts(plot=dict(show_legend=False))
plot = datashade(overlay, aggregator=ds.count_cat('types'), color_key=color_key)

hover = HoverTool(
    tooltips=[
        ("Date", "$x{%F %T}"),
        ("Value", "$y{s}"),
    ],
    formatters={
        'x'      : 'datetime', 
    },
)
plot_opts = dict(width=800, height=300, tools=[hover])
plot.opts(plot=plot_opts)
```

**Comment by karlglazebrook** [comment]
> I think this leaves the door open for custom formatter models in the future, with something like
> 
> ```
> HoverTool(tooltips=[("date", "@x{custom}")], formatters={"x": my_custom_formatter})
> ```
> 

This is nice and I found it in a google search on how to do this. Thanks! 

But I was wondering - is the use of "@x" (as opposed to "$x") documented anyway? I searched the docs and couldn't find it. It seems to give the actual point value as opposed to "the data value nearby where the mouse is." I get the impression that @x and @y are provided as some kind of default ColumnDataSource.
